### ISSUE TITLE
Add Android Auto connection detection to trigger options for High Accuracy mode

### ISSUE THREAD

**Issue by crowedavid** [home-assistant/android#1516]
**Is your feature request related to a problem? Please describe.**

Currently, the BT connected device option works well detecting an Android Auto wired (USB) connection since it retains the BT connection in Android Auto mode. Soon I will have an Android Auto [device](https://www.indiegogo.com/projects/aawireless#/) allowing me to use a wireless connection (generally using WiFi Direct but also over an in car WiFi hotspot in the future) that will drop the BT connection when it connects to the AA headunit. This is normal behavior for AA when it uses a wireless connection.

**Describe the solution you'd like**

Add an option to add Android Auto connection (independent of BT or wireless) detection as an option to trigger High Accuracy mode.

**Describe alternatives you've considered, if any**

Alternative for short term is using another zone to trigger high accuracy mode but this complicates trigger timing so it will arm/disarm correctly.

**Additional context**

**Comment by dshokouhi** [comment]
I imagine the phone will report a different connected SSID for this hopefully it wouldn't be hidden to the app.  If you see the wifi connection sensor updating to the connected SSID then this would be doable otherwise I wonder if google will have it hidden from non android auto apps.

**Comment by crowedavid** [comment]
i should have my new device in the next few weeks (still in distribution from factory) and will have more specific information on the connection sequence.  i'm not an Android dev but was hoping there might be a programatic way to obtain the Android Auto mode switch on/off. it does initiate a couple of persistent notifications when it starts and then disappears when disconnected.

**Comment by crowedavid** [comment]
i received my AAWireless puck and it works great (with the latest firmware upgrade).  with my headunit a Bluetooth connection is available to use to trigger the High Accuracy mode.

with that said, it would still be useful (but not critical) to have the Home Assistant companion app detect when Android Auto mode is active. Tasker does detect it and changes the built-in variable %UIMODE from "normal" to "car" which can be used as a state change to trigger a task sending an intent.

i suppose one option might be for the HA companion app to expose an intent target to send a message to in order to trigger the High Accuracy mode on/off when needed.  another would be to incorporate the mechanism Tasker is using to detect the change.

thanks for your consideration.

**Comment by dshokouhi** [comment]
the android app can already listen to and send intents for any intent, you just need to know what the intent is then you can automate off of it.

**Comment by crowedavid** [comment]
i do understand i can listen to an intent in the android app and send an action to my home assistant instance.  unless i am missing something, it would seem i would have to have the android app receive an intent that might initiate an automation that would send a notification back containing the trigger to turn High Accuracy mode on/off.  i don't really want to turn the entire function on/off with this - i just want to have Android Auto trigger it instead of BT devices.

currently, the High Accuracy mode can be triggered by a list of BT devices and/or a list of zones. my idea is to have the option of triggering it by presence of an Android Auto connection.

**Comment by Mincka** [comment]
Would it be possible to use the `CarConnection` API ([documented here](https://developer.android.com/training/cars/apps#car-connection)) to add a sensor to the app?

It should work for Android Auto over USB and Wi-Fi.
With this kind of sensor, the user could use an automation to use it as a trigger or a condition.

In my case, I would find it useful to use it as a condition to open the gate when I am coming home in the car, as an additional control with location.

It's more precise than activity detection and not based on car Wi-FI AP / BT detection (that can be also disabled, not configured...).

**Comment by dshokouhi** [comment]
> Would it be possible to use the `CarConnection` API ([documented here](https://developer.android.com/training/cars/apps#car-connection)) to add a sensor to the app?

This API is only for android auto apps which we are not, needs to be in #791

**Comment by Mincka** [comment]
Ok, I was not sure about this based on [this comment](https://stackoverflow.com/questions/70051192/how-to-detect-android-auto-is-connected-with-android-12#comment126203680_71384066).

So you say that the Car App Library is not available at all if the application is not an "Android Auto App"?

**Comment by dshokouhi** [comment]
Before we dive in, we are going off track on the topic of this issue lets move to a new one as this issue is not asking for a sensor like you did.

**Comment by Mincka** [comment]
Sure, my initial intention was to suggest a possible alternative to avoid cluttering the High Accuracy sensor in favor of a more general solution.
I can create a new issue since a new sensor does not fit exactly in this one or #791.

**Comment by dshokouhi** [comment]
#791 is for the app to be added to the android auto store, your ask is to add a sensor to the phone app so thats different. 

I do like the idea of exposing it as a sensor so users can do their own automations to prevent high accuracy mode from getting too complicated.

**Comment by GSzabados** [comment]
As #791 has been closed this might be possible as well.

**Comment by crowedavid** [comment]
i am just catching up on the recent Android Auto beta and as the original requester would suggest the intent for this issue isn't fixed by #3228.

i do appreciate the Android Auto addition so far but would like to have the Android Auto connected sensor to be used in the High Accuracy location mode in parallel to the BT connection status.  specifically, use the list of BT devices or Android Auto detected in conjunction with the entering zone setting.

logically, like this: (BT device OR Android Auto) AND entering zone

this simplifies the BT device list significantly for me. thanks.

**Comment by dshokouhi** [comment]
> logically, like this: (BT device OR Android Auto) AND entering zone

This is the reason why we have a new sensor and the notification command so you can create your own logic. If we add too much logic here it begins to get complex. Some users may start to request when their device is charging or when a sensor changes etc... For that reason its best to leave the update logic in the hands of the user when things get more complex than the current settings.

**Comment by crowedavid** [comment]
> This is the reason why we have a new sensor and the notification command so you can create your own logic. If we add too much logic here it begins to get complex. Some users may start to request when their device is charging or when a sensor changes etc... For that reason its best to leave the update logic in the hands of the user when things get more complex than the current settings.

my request is for on device specific logic that is responding to and influencing local conditions only on the device itself. the nature and use of detecting Android Auto is very similar to that of a BT connection.

you're suggesting either use the current on device logic for BT+zone or, if i also want to add Android Auto detection to the mix, rip that all out and just force HA mode on or off via automation->notifications.  that's a hammer approach when a more finessed approach appears to be reasonable.

if yours is the final word, closing this issue as "fixed" is incorrect. if you aren't going to entertain the option set it with "closed - won't fix" so it reflects the dev team stance more correctly.

**Comment by Mincka** [comment]
Just curious, what's your use case for enabling High Accuracy when the phone is connected to Android Auto?

In my case, since I use Waze / Maps in AA, the location is already sent through HA app every few seconds with the general sensors update loop.

Having High Accuracy for the whole duration of your car trip will be a huge battery drain (especially with AAWireless) with limited added accuracy.

**Comment by crowedavid** [comment]
> Just curious, what's your use case for enabling High Accuracy when the phone is connected to Android Auto?

android auto detection and BT connection detection serve the same purpose for me: turn on high accuracy when i'm in a vehicle and close to my home where i want to detect whether i'm in my home zone before i reach home. the current logic in the companion app works well with BT and zone detection. i just want to add the ability to have the same detection available for android auto only so it frees from maintaining the BT connection list.

> In my case, since I use Waze / Maps in AA, the location is already sent through HA app every few seconds with the general sensors update loop.
> 
> Having High Accuracy for the whole duration of your car trip will be a huge battery drain (especially with AAWireless) with limited added accuracy.

i don't have high accuracy enabled for the duration and that's really the main point. i would like the on device state to trigger high accuracy for the short period of time when i'm in a vehicle moving fast enough that the normal location updates don't reliably detect home arrival in time.

**Comment by GSzabados** [comment]
> i don't have high accuracy enabled for the duration and that's really the main point. i would like the on device state to trigger high accuracy for the short period of time when i'm in a vehicle moving fast enough that the normal location updates don't reliably detect home arrival in time.

I don't know where you live and what speed are you driving your car that you would have location issues, but high accuracy mode is really for areas where are high buildings and GPS antennas would have issues to receive signal from the GPS satellites. Or you are driving underground in a parking lot where are WiFi antennas which could support positioning.

High accuracy mode turns on Bluetooth and WiFi support. And works well where are fixed WiFi APs what has been mapped by Google. It would not really give any advantage over arrival time. If your problem is update, then you should really look at your internet connections on your phone.

**Comment by crowedavid** [comment]
> > i don't have high accuracy enabled for the duration and that's really the main point. i would like the on device state to trigger high accuracy for the short period of time when i'm in a vehicle moving fast enough that the normal location updates don't reliably detect home arrival in time.
> 
> I don't know where you live and what speed are you driving your car that you would have location issues, but high accuracy mode is really for areas where are high buildings and GPS antennas would have issues to receive signal from the GPS satellites. Or you are driving underground in a parking lot where are WiFi antennas which could support positioning.
> 
> High accuracy mode turns on Bluetooth and WiFi support. And works well where are fixed WiFi APs what has been mapped by Google. It would not really give any advantage over arrival time. If your problem is update, then you should really look at your internet connections on your phone.

thanks for offering your intuitions on my use case.  as i noted, without high accuracy mode the updates are not reliable enough to detect when i'm in my home zone. with high accuracy mode they work well.

**Comment by GSzabados** [comment]
How big is your home zone?

**Comment by crowedavid** [comment]
home zone is 130m and the approach radius i use in the companion app is 800m to trigger high accuracy mode. with my neighborhood arrangement and use patterns this works well for my needs for detecting when i'm home vs at the park or visiting friends nearby.

**Comment by dshokouhi** [comment]
> you're suggesting either use the current on device logic for BT+zone or, if i also want to add Android Auto detection to the mix, rip that all out and just force HA mode on or off via automation->notifications. that's a hammer approach when a more finessed approach appears to be reasonable.

You actually don't need to rip it all out. We have 2 different sets of commands for high accuracy. One allows you to simply turn it on and to continue to respect the BT and Zone logic. The other allows you to override those changes until the next condition change. This allows you to keep the app logic and to add your own conditions. We also have sensors that allow you to understand if high accuracy mode is on and the state of BT connection.

https://companion.home-assistant.io/docs/notifications/notification-commands#high-accuracy-mode



> if yours is the final word, closing this issue as "fixed" is incorrect. if you aren't going to entertain the option set it with "closed - won't fix" so it reflects the dev team stance more correctly.

This is actually an alternative option which allows us to keep things open so users can add their own specific logic without us having to add more options and settings. As we cannot account for all the different use cases a user may have to detect when they are in their vehicle this is the best of both worlds.  I understand it may create more work for your own use cases but it allows us to be adaptable to all use cases.

> In my case, since I use Waze / Maps in AA, the location is already sent through HA app every few seconds with the general sensors update loop.

correct when another app is using GPS we will get passive updates every 30 seconds without needing to enter high accuracy mode.

**Comment by Mincka** [comment]
> i don't have high accuracy enabled for the duration and that's really the main point. **i would like the on device state to trigger high accuracy for the short period of time** when i'm in a vehicle moving fast enough that the normal location updates don't reliably detect home arrival in time.

So you are saying that you don't want HA "on" when AA is "on", but just when you are coming back home, for a "short period of time".
It's even more complicated to implement locally because it means it would require a combination of conditions.
If "AA "on" AND Approach Zone "on" then HA "on".

I'm not saying that's a bad idea but there's an history for this sensor and each iteration makes it more complicated to maintain for the developers.
There's already a switch in the App to force BT AND Zone together, otherwise I guess it's a OR.

All the current options in the HA sensor came from good intentions but it ultimately leads to this kind of issue: someone else wants another condition and the whole AND / OR logic in this sensor cannot be maintained. Some people might want (BT AND Zone) OR (AA AND Zone), there's no easy solution. Either the devs choose to stop the bleeding there or it would require a whole new logic that could accommodate all cases but would not be user friendly and without breaking changes.

So comparing this with you just needing to send a notification when the trigger is entering the "Approaching Zone" with condition "AA is "on", I don't think it can be called "hammer approach".

**Comment by crowedavid** [comment]
> > you're suggesting either use the current on device logic for BT+zone or, if i also want to add Android Auto detection to the mix, rip that all out and just force HA mode on or off via automation->notifications. that's a hammer approach when a more finessed approach appears to be reasonable.
> 
> You actually don't need to rip it all out. We have 2 different sets of commands for high accuracy. One allows you to simply turn it on and to continue to respect the BT and Zone logic. The other allows you to override those changes until the next condition change. This allows you to keep the app logic and to add your own conditions. We also have sensors that allow you to understand if high accuracy mode is on and the state of BT connection.
> 
> https://companion.home-assistant.io/docs/notifications/notification-commands#high-accuracy-mode

thanks for your suggestion and pointing to the additional high accuracy mode commands you've implemented.

i spent a bit of time implementing and testing the android auto sensor triggered notification automations based setup again and was reminded of what frustrated me most with this approach the last time: it depends entirely on certainty and timeliness in delivering notifications to my android handset (pixel 7 w/ latest firmware and companion app beta) for the commands to take effect. the issue becomes that turning off the high accuracy mode occurs at the exact time when my handset is transitioning from cell to home network and the seconds that takes will frequently be in the exact timespan the sensor->home assistant host->notification->high accuracy mode force-off command and the notification gets lost in the shuffle.

i currently have "persistent connection" set to "on home wifi only" so that may contribute to the timing and loss of the notification with the command. i haven't tested with the "persistent connection" on another option but do know i lost notifications with the google only option ("never") during this transition time frequently before the new persistent options were available.

if this portion of the logic was on the handset only i wouldn't be playing roulette with lost notifications to receive commands.

during testing i believe i have a different issue that cropped up that i'll try to get a repeatable sequence on to create a new issue.

**Comment by crowedavid** [comment]
> > i don't have high accuracy enabled for the duration and that's really the main point. **i would like the on device state to trigger high accuracy for the short period of time** when i'm in a vehicle moving fast enough that the normal location updates don't reliably detect home arrival in time.
> 
> So you are saying that you don't want HA "on" when AA is "on", but just when you are coming back home, for a "short period of time". It's even more complicated to implement locally because it means it would require a combination of conditions. If "AA "on" AND Approach Zone "on" then HA "on".
> 
> I'm not saying that's a bad idea but there's an history for this sensor and each iteration makes it more complicated to maintain for the developers. There's already a switch in the App to force BT AND Zone together, otherwise I guess it's a OR.
> 
> All the current options in the HA sensor came from good intentions but it ultimately leads to this kind of issue: someone else wants another condition and the whole AND / OR logic in this sensor cannot be maintained. Some people might want (BT AND Zone) OR (AA AND Zone), there's no easy solution. Either the devs choose to stop the bleeding there or it would require a whole new logic that could accommodate all cases but would not be user friendly and without breaking changes.
> 
> So comparing this with you just needing to send a notification when the trigger is entering the "Approaching Zone" with condition "AA is "on", I don't think it can be called "hammer approach".

the current code implements the following:

```
BT device AND entering zone (with approach radius)
```

what i'm asking for is:

```
(Android Auto OR BT device) AND entering zone (with approach radius)
```

that's it. see my other issues with the notification based commands in my comment just [above](https://github.com/home-assistant/android/issues/1516#issuecomment-1407179337).

**Comment by dshokouhi** [comment]
> i spent a bit of time implementing and testing the android auto sensor triggered notification automations based setup again and was reminded of what frustrated me most with this approach the last time: it depends entirely on certainty and timeliness in delivering notifications to my android handset (pixel 7 w/ latest firmware and companion app beta) for the commands to take effect. the issue becomes that turning off the high accuracy mode occurs at the exact time when my handset is transitioning from cell to home network and the seconds that takes will frequently be in the exact timespan the sensor->home assistant host->notification->high accuracy mode force-off command and the notification gets lost in the shuffle.
> 
> i currently have "persistent connection" set to "on home wifi only" so that may contribute to the timing and loss of the notification with the command. i haven't tested with the "persistent connection" on another option but do know i lost notifications with the google only option ("never") during this transition time frequently before the new persistent options were available.
> 
> if this portion of the logic was on the handset only i wouldn't be playing roulette with lost notifications to receive commands.

As you are using the google play version there is a FCM fallback. I am not sure what format of the command you are using but this behavior you describe about a lost notification sounds like you may not be using the [critical notification format](https://companion.home-assistant.io/docs/notifications/critical-notifications#android). So even if you are using the persistent connection, notifications may still arrive via FCM when the connection is not established. You don't need this format with persistent connection but you do want it for notifications that need to be delivered immediately via firebase.

The issue isn't that the notification is lost its that by default FCM notifications respect doze mode and as such the notification will not be delivered until the next time the device wakes up. They appear lost or moreso delayed. One thing you can do is you can use the [notification received event](https://companion.home-assistant.io/docs/notifications/notification-received) to determine when a particular notification was received by the device.

There is also a developer option in Android for "Mobile Data Always Active" I am pretty sure its on by default but may be worthwhile to check to keep the wifi >cell transition faster.

**Comment by Mincka** [comment]
> what i'm asking for is:
> 
> ```
> (Android Auto OR BT device) AND entering zone (with approach radius)
> ```

Thanks for confirming exactly what I was saying in my post. You want your own logic combination that is complicated to implement in the local logic of the App. Why not also asking for "Wi-Fi connection to trigger High Accuracy"? Another condition with AND / OR logic that someone would certainly want.

You can make your own infinite combination of conditions on the entering zone trigger in the automation. Here's what I use:
![image](https://user-images.githubusercontent.com/6184289/215255373-23d0c2c0-7c36-45c5-b29d-97bd147edfcf.png)

Regarding the HA notification not working, I use critical notification format + repeat just in case:

```
action:
  - choose:
      - conditions:
          - condition: trigger
            id: julien_enable_high_accuracy
          - condition: state
            entity_id: binary_sensor.telephone_julien_high_accuracy_mode
            state: "off"
        sequence:
          - repeat:
              until:
                - condition: template
                  value_template: >-
                    {{
                    is_state('binary_sensor.telephone_julien_high_accuracy_mode',
                    'on') or repeat.index > 4 }}
              sequence:
                - service: notify.mobile_app_telephone_julien
                  data:
                    message: command_high_accuracy_mode
                    data:
                      command: turn_on
                      ttl: 0
                      priority: high
                - delay:
                    hours: 0
                    minutes: 0
                    seconds: 30
                    milliseconds: 0
```

To be on your side for the original request, I think the only way to satisfy all possible cases locally would be to have a Jinja template in app sensor for High Accuracy activation. Using Jinja templates in the Apps is something already exists elsewhere, in the widgets on Android and on in the Complications for the Apple Watch. It could satisfy power users for custom cases.
